docs/infra/example_lightning.py

```diff
@@ -0,0 +1,155 @@
+    def training_step(self, batch, batch_idx):
+        loss = self._step(batch)
+        self.log("train_loss", loss)
+        return loss
+
+    def validation_step(self, batch, batch_idx):
+        loss = self._step(batch)
+        self.log("val_loss", loss)
+        return loss
+
+    def configure_optimizers(self):
+        return torch.optim.Adam(self.model.parameters(), lr=self.learning_rate)
+
+
+class Data(pl.LightningDataModule):
+    def __init__(self, batch_size: int):
+        super().__init__()
+        self.batch_size = batch_size
+
+    def _dataloader(self, train: bool):
+        transform = transforms.Compose([transforms.ToTensor(), transforms.Normalize((0.5,), (0.5,))])
+        dset = datasets.MNIST('', train=train, download=True, transform=transform)
+        return torch.utils.data.DataLoader(dset, batch_size=self.batch_size)
+
+    def train_dataloader(self):
+        return self._dataloader(train=True)
+
+    def val_dataloader(self):
+        return self._dataloader(train=False)
+    
+
+
+class ModelConfig(pydantic.BaseModel):
+    pretrained: bool = True
+    learning_rate: float = 0.001
+    
+    model_config = pydantic.ConfigDict(extra="forbid")
+
+    def build(self) -> Model:
+        return Model(**self.dict())
+
+
+class DataConfig(pydantic.BaseModel):
+    batch_size: int = 64
+    
+    model_config = pydantic.ConfigDict(extra="forbid")
+
+    def build(self) -> Data:
+        return Data(**self.dict())
+
+
+class TrainerConfig(pydantic.BaseModel):
+    max_epochs: tp.Optional[int] = None
+    
+    model_config = pydantic.ConfigDict(extra="forbid")
+
+    def build(self, checkpoint_path: str | None = None) -> Trainer:
+        if checkpoint_path:
+            callbacks = [ModelCheckpoint(
+                dirpath=checkpoint_path,
+                save_top_k=1,
+                monitor="val_loss",
+                mode="min")
+            ]
+        else:
+            callbacks = None
+        return Trainer(**self.dict(), callbacks=callbacks)
+
+class Experiment(pydantic.BaseModel):
+    model: ModelConfig = ModelConfig()
+    data: DataConfig = DataConfig()
+    trainer: TrainerConfig = TrainerConfig()
+    infra: exca.TaskInfra = exca.TaskInfra(folder='.cache/')
+
+    @property
+    def checkpoint_path(self):
+        # Define the checkpoint directory
+        checkpoint_dir = self.infra.uid_folder() / 'checkpoint'
+
+        # Find the latest checkpoint if it exists
+        checkpoints = sorted(checkpoint_dir.glob('*.ckpt'))
+        ckpt_path = sorted(checkpoints)[-1] if checkpoints else None
+        return ckpt_path
+
+    @infra.apply
+    def fit(self):
+        # Configure
+        data = self.data.build()
+        model = self.model.build()
+        trainer = self.trainer.build(self.infra.folder)
+
+        # Fit model
+        trainer.fit(model, data, ckpt_path=self.checkpoint_path)
+        
+        # Return model if not saved 
+        if self.checkpoint_path is None:
+            return model
```

Comment:
this is solving the caching issue of models with a really bad practice (changing output type depending on the inputs) :s I don't know how to solve this one though